File: domusviewer/app/src/main/java/it/achdjian/paolo/domusviewer/DomusEngineRest/ConnectionStatus.java
package it.achdjian.paolo.domusviewer.DomusEngineRest;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by paolo on 14/04/16.
 */
public class ConnectionStatus {
    private boolean connected=false;
    private List<ConnectionObserver> observers= new ArrayList<>();

    public boolean isConnected(){
        return connected;
    }

    public void setConnected(boolean connected){
        if(this.connected != connected) {
            this.connected = connected;
            for (ConnectionObserver observer : observers) {
                if (connected) {
                    observer.connected();
                } else {
                    observer.disconnected();
                }
            }
        }
    }

    public void addObserver(ConnectionObserver observer){
        observers.add(observer);
    }

    public void removeObserver(ConnectionObserver observer){
        observers.remove(observer);
    }


}

File: temperature_monitor/app/src/main/java/it/achdjian/paolo/temperaturemonitor/domusEngine/ConnectionStatus.kt
package it.achdjian.paolo.temperaturemonitor.domusEngine

import android.util.Log
import it.achdjian.paolo.temperaturemonitor.domusEngine.rest.ZigbeeRunnable
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.properties.Delegates

/**
 * Created by paolo on 14/04/16.
 */
@Singleton
class ConnectionStatus @Inject constructor(){
    private val observers = ArrayList<ConnectionObserver>()

    var connected: Boolean
            by Delegates.observable(false) {
                _, _, new ->
                Log.i(ZigbeeRunnable.TAG, "set new value: " + new)
                Log.i(ZigbeeRunnable.TAG, observers.size.toString() + " observers")
                for (observer in observers) {
                    if (new)
                        observer.connected()
                    else
                        observer.disconnected()
                }
            }

    fun addObserver(observer: ConnectionObserver) {
        Log.i(ZigbeeRunnable.TAG,"add observer")
        observers.add(observer)
    }

    fun removeObserver(observer: ConnectionObserver) {
        observers.remove(observer)
    }


}
